spark-bigtable_2.12/src/main/scala/com/google/cloud/spark/bigtable/datasources/BigtableTableScanRDD.scala

```diff
@@ -0,0 +1,229 @@
+import com.google.protobuf.ByteString
+import org.apache.spark.rdd.RDD
+import org.apache.spark.{Partition, TaskContext}
+import org.apache.yetus.audience.InterfaceAudience
+
+import scala.collection.JavaConverters._
+
+@InterfaceAudience.Private
+class BigtableTableScanRDD(
+    relation: BigtableRelation,
+    clientKey: BigtableClientKey,
+    filterRangeSet: RangeSet[RowKeyWrapper]
+) extends RDD[BigtableRow](relation.sqlContext.sparkContext, Nil) {
+
+  override def getPartitions: Array[Partition] = {
+    try {
+      val clientHandle = BigtableDataClientBuilder.getHandle(clientKey)
+      val bigtableDataClient = clientHandle.getClient()
+
+      val keyOffsets: List[KeyOffset] =
+        bigtableDataClient.sampleRowKeys(relation.tableId).asScala.toList
+
+      val tabletRanges: Array[BigtableTabletRange] =
+        new Array[BigtableTabletRange](keyOffsets.size)
+      // We use empty ByteString as the smallest row key and null as the largest.
+      var start: ByteString = ByteString.EMPTY
+      var end: ByteString = null
+      var i = 0
+      while (i < keyOffsets.size - 1) {
+        end = keyOffsets(i).getKey
+        tabletRanges(i) = BigtableTabletRange(i, start, end)
+        start = end
+        i += 1
+      }
+      // Ensure that the last tablet's 'end key' is considered null since
+      //  it's considered the maximum value inside RowKeyWrapper.
+      tabletRanges(i) = BigtableTabletRange(i, start, null)
+
+      // Consider switching to using shard(List<KeyOffset> sampledRowKeys) from
+      //  com.google.cloud.bigtable.data.v2.models.Query for intersecting
+      //  tablet ranges with filters.
+      var idx = 0
+      val partitions = tabletRanges.flatMap { tabletRange =>
+        val partitionRangeSet: RangeSet[RowKeyWrapper] =
+          TreeRangeSet.create[RowKeyWrapper]()
+        partitionRangeSet.add(
+          GuavaRange.closedOpen(
+            new RowKeyWrapper(tabletRange.start),
+            new RowKeyWrapper(tabletRange.end)
+          )
+        )
+        partitionRangeSet.removeAll(filterRangeSet.complement())
+        if (!partitionRangeSet.isEmpty) {
+          idx += 1
+          Some(
+            BigtableScanPartition(
+              idx - 1,
+              tabletRange,
+              partitionRangeSet
+            )
+          )
+        } else {
+          None
+        }
+      }
+
+      clientHandle.close()
+      partitions.asInstanceOf[Array[Partition]]
+    } catch {
+      case e: Exception => {
+        logError("Received error when creating partitions: " + e.getMessage)
+        throw e
+      }
+    }
+  }
+
+  private def streamToIterator(
+      stream: ServerStream[BigtableRow],
+      clientHandle: BigtableDataClientBuilder.DataClientHandle
+  ): Iterator[BigtableRow] = {
+    val it = stream.iterator()
+    val iterator = new Iterator[BigtableRow] {
+      override def hasNext: Boolean = {
+        if (!it.hasNext) {
+          stream.cancel()
```

Comment:
I don't think we need to cancel the stream since we've read everything?

Reply:
Since we don't have automated tests running for the GitHub repo yet, I think might be safer to publish the first version with minimal logical changes compared to what we've already tested, but I added a TODO to consider removing this after merging the first iteration.